docs/4-functional-programming/fp-hometask/src/App.tsx

```diff
@@ -0,0 +1,78 @@
+import { FC } from 'react';
+import { useState, useEffect } from 'react';
+import { StyledEngineProvider } from '@mui/material/styles';
+
+import { Image, User, Account } from '../types';
+import { Table, Filters, Sort, Search, Row } from './components';
+import { getImages, getUsers, getAccounts } from './mocks/api';
+import rows from './mocks/rows.json';
+import styles from './App.module.scss';
+import { dataConverter } from './dataConverter';
+import {
+  filterByFilters,
+  filterBySearch,
+  getUniqueRows,
+  sortRows,
+} from './utils';
+import { FilterValueType, SortValue } from './types';
+
+// mockedData has to be replaced with parsed Promises’ data
+const mockedData: Row[] = rows.data;
+
+let ALL_ROWS: Row[] = [];
+
+export const App: FC = () => {
+  const [data, setData] = useState<Row[]>(undefined);
+  const [searchedValue, setSearchedValue] = useState('');
+  const [selectedFilters, setSelectedFilters] = useState<FilterValueType[]>([]);
+  const [selectedSort, setSelectedSort] = useState<SortValue>('');
+
+  useEffect(() => {
+    // fetching data from API
+    Promise.all([getImages(), getUsers(), getAccounts()]).then(
+      ([images, users, accounts]: [Image[], User[], Account[]]) => {
+        const convertedData = dataConverter(users, accounts, images);
+        ALL_ROWS = convertedData;
+        setData(convertedData);
+      }
+    );
+  }, []);
+
+  useEffect(() => {
+    const filteredData = [
+      ...filterBySearch(ALL_ROWS, searchedValue),
+      ...filterByFilters(ALL_ROWS, selectedFilters),
+    ];
+
+    let newData: Row[] = [];
+
+    if (searchedValue || selectedFilters.length) {
+      newData = getUniqueRows(filteredData);
+    } else {
+      newData = ALL_ROWS;
+    }
+
+    setData(sortRows(newData, selectedSort));
+  }, [searchedValue, selectedFilters, selectedSort]);
```

Comment:
We want to use something like:

```
setData(sortRows(filterRows(ALL_ROWS)));
```

Reply:
for appropriate filtering and sorting we need sortValue + value from search + all selected filters. It's obvious. We can put filterRows and sortRows in utils to hide them and for to extend our component or we can put these two functions in component then we can pass only ALL_ROWS and every internal function will take required value from the component state

Reply:
in FP the idea should look something like this:

```
const rowSorter = curry(sortRows)(selectedSort); // order of the sortRows arguments should be reversed
const rowFilter = curry(filterRows)(searchedValue)(selectedFilters); // order of the filterRows arguments should be updated
const dataMutator = compose(rowFilter, rowSorter);
setDataForView(dataMutator(data));
// OR
// setDataForView(rowSorter(rowFilter(data)))

```

Reply:
I hope I implemented correctly